pageserver/src/tenant/layer_map.rs

```diff
@@ -63,6 +64,22 @@ pub use historic_layer_coverage::Replacement;
 
 use super::storage_layer::range_eq;
 
+struct LayerRef<L: ?Sized>(Arc<L>);
+
+impl<L: ?Sized> PartialEq for LayerRef<L> {
+    fn eq(&self, other: &LayerRef<L>) -> bool {
+        Arc::ptr_eq(&self.0, &other.0)
+    }
+}
+
+impl<L: ?Sized> Eq for LayerRef<L> {}
+
+impl<L: ?Sized> Hash for LayerRef<L> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        Arc::into_raw(Arc::clone(&self.0)).hash(state)
```

Comment:
```suggestion
        Arc::as_ptr(&self.0).hash(state)
```

This will avoid incrementing and decrementing the strong counter needlessly around hashing. No comment on if this makes sense ... With the current `Arc::ptr_eq` it might be in violation of the trait if used together with `Eq`. I think I added more comments on that in `LayerMap::compared_arced_layers`.

Reply:
This would had been a memory leak with `into_raw`.